app/src/main/java/org/openobservatory/ooniprobe/activity/runtests/RunTestsActivity.kt

```diff
@@ -0,0 +1,189 @@
+package org.openobservatory.ooniprobe.activity.runtests
+
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import android.text.TextUtils
+import android.view.*
+import org.openobservatory.ooniprobe.R
+import org.openobservatory.ooniprobe.activity.AbstractActivity
+import org.openobservatory.ooniprobe.activity.RunningActivity
+import org.openobservatory.ooniprobe.activity.runtests.RunTestsViewModel.Companion.SELECT_ALL
+import org.openobservatory.ooniprobe.activity.runtests.RunTestsViewModel.Companion.SELECT_NONE
+import org.openobservatory.ooniprobe.activity.runtests.RunTestsViewModel.Companion.SELECT_SOME
+import org.openobservatory.ooniprobe.activity.runtests.adapter.RunTestsExpandableListViewAdapter
+import org.openobservatory.ooniprobe.activity.runtests.models.ChildItem
+import org.openobservatory.ooniprobe.activity.runtests.models.GroupItem
+import org.openobservatory.ooniprobe.common.PreferenceManager
+import org.openobservatory.ooniprobe.databinding.ActivityRunTestsBinding
+import org.openobservatory.ooniprobe.test.suite.*
+import java.io.Serializable
+import javax.inject.Inject
+
+class RunTestsActivity : AbstractActivity() {
+	lateinit var binding: ActivityRunTestsBinding
+
+	private lateinit var mAdapter: RunTestsExpandableListViewAdapter
+
+	@Inject
+	lateinit var preferenceManager: PreferenceManager
+
+	@Inject
+	lateinit var viewModel: RunTestsViewModel
+
+	companion object {
+		const val TESTS: String = "tests"
+
+		@JvmStatic
+		fun newIntent(context: Context, testSuites: List<AbstractSuite>): Intent {
+			return Intent(context, RunTestsActivity::class.java).putExtras(Bundle().apply {
+				putSerializable(TESTS, testSuites as Serializable)
+			})
+		}
+	}
+
+	override fun onCreate(savedInstanceState: Bundle?) {
+		super.onCreate(savedInstanceState)
+		binding = ActivityRunTestsBinding.inflate(layoutInflater)
+		setContentView(binding.getRoot())
+
+		setSupportActionBar(binding.toolbar)
+		supportActionBar?.setDisplayHomeAsUpEnabled(true)
+
+		activityComponent?.inject(this)
+
+		val testSuites: List<AbstractSuite>? = intent.extras?.getParcelableArrayList(TESTS)
+		testSuites?.let { ts ->
+			val tsGroups: List<GroupItem> = ts.map { testSuite ->
+				return@map testSuite.runTestsGroupItem(preferenceManager)
+			}
+
+			mAdapter = RunTestsExpandableListViewAdapter(this, tsGroups, viewModel)
+
+			binding.expandableListView.setAdapter(mAdapter)
+			for (i in 0 until mAdapter.groupCount) {
+				binding.expandableListView.expandGroup(i)
+			}
+			binding.selectAll.setOnClickListener { onSelectAllClickListener() }
+
+			binding.selectNone.setOnClickListener { onSelectNoneClickListener() }
+
+			viewModel.selectedAllBtnStatus.observe(this, this::selectAllBtnStatusObserver)
+
+			binding.bottomBar.setOnMenuItemClickListener { menuItem ->
+				onMenuItemClickListener(menuItem)
+			}
+		} ?: run {
```

Comment:
What does this do? I don't fully understand this LoC

Reply:
Here's a breakdown of the expression:

- `testSuites.let {}` : This checks if the object is not null. If not null, it calls the lambda expression inside the curly braces with the object as the argument (referred to as it).
- `?:` : The Elvis operator acts as a fallback mechanism.
- `run {}` : If the object is null, the run lambda block is executed instead. This block can contain alternative actions for when the object is null.